File: bindings/python/satox_bindings/security_manager.py
"""
Satox SDK Security Manager Binding
"""

from typing import Dict, List, Optional, Any
import asyncio
from .error import SatoxError

class SecurityManager:
    """Wrapper for the C++ SecurityManager"""
    
    def __init__(self):
        self._core = None
        self._initialized = False
        self._callbacks = []
    
    async def initialize(self, config: Dict[str, Any]) -> bool:
        """Initialize security manager"""
        if self._initialized:
            raise SatoxError("SecurityManager already initialized")
        
        try:
            result = await self._call_native("initialize", config)
            self._initialized = result
            return result
        except Exception as e:
            raise SatoxError(f"Failed to initialize SecurityManager: {str(e)}")
    
    async def generate_key_pair(self) -> Dict[str, str]:
        """Generate key pair"""
        self._ensure_initialized()
        
        try:
            return await self._call_native("generate_key_pair")
        except Exception as e:
            raise SatoxError(f"Failed to generate key pair: {str(e)}")
    
    async def encrypt_data(self, data: bytes, public_key: str) -> bytes:
        """Encrypt data"""
        self._ensure_initialized()
        
        try:
            return await self._call_native("encrypt_data", data, public_key)
        except Exception as e:
            raise SatoxError(f"Failed to encrypt data: {str(e)}")
    
    async def decrypt_data(self, encrypted_data: bytes, private_key: str) -> bytes:
        """Decrypt data"""
        self._ensure_initialized()
        
        try:
            return await self._call_native("decrypt_data", encrypted_data, private_key)
        except Exception as e:
            raise SatoxError(f"Failed to decrypt data: {str(e)}")
    
    async def sign_data(self, data: bytes, private_key: str) -> bytes:
        """Sign data"""
        self._ensure_initialized()
        
        try:
            return await self._call_native("sign_data", data, private_key)
        except Exception as e:
            raise SatoxError(f"Failed to sign data: {str(e)}")
    
    async def verify_signature(self, data: bytes, signature: bytes, public_key: str) -> bool:
        """Verify signature"""
        self._ensure_initialized()
        
        try:
            return await self._call_native("verify_signature", data, signature, public_key)
        except Exception as e:
            raise SatoxError(f"Failed to verify signature: {str(e)}")
    
    def _ensure_initialized(self):
        if not self._initialized:
            raise SatoxError("SecurityManager not initialized")
    
    async def _call_native(self, method: str, *args) -> Any:
        """Call native C++ method"""
        raise NotImplementedError("Native binding not implemented") 

File: bindings/typescript/src/security-manager.ts
/**
 * Satox SDK Security Manager Binding
 */

import { SatoxError } from './error';

export class SecurityManager {
    private core: any;
    private initialized: boolean = false;
    private callbacks: Array<(event: string, data: any) => void> = [];

    constructor() {
        this.core = require('./native/satox_core');
    }

    async initialize(config: Record<string, any>): Promise<boolean> {
        if (this.initialized) {
            throw new SatoxError('SecurityManager already initialized');
        }

        try {
            const result = await this.core.securityInitialize(config);
            this.initialized = result;
            return result;
        } catch (error) {
            throw new SatoxError(`Failed to initialize SecurityManager: ${error.message}`);
        }
    }

    async generateKeyPair(): Promise<Record<string, string>> {
        this.ensureInitialized();

        try {
            return await this.core.securityGenerateKeyPair();
        } catch (error) {
            throw new SatoxError(`Failed to generate key pair: ${error.message}`);
        }
    }

    async encryptData(data: Buffer, publicKey: string): Promise<Buffer> {
        this.ensureInitialized();

        try {
            return await this.core.securityEncryptData(data, publicKey);
        } catch (error) {
            throw new SatoxError(`Failed to encrypt data: ${error.message}`);
        }
    }

    async decryptData(encryptedData: Buffer, privateKey: string): Promise<Buffer> {
        this.ensureInitialized();

        try {
            return await this.core.securityDecryptData(encryptedData, privateKey);
        } catch (error) {
            throw new SatoxError(`Failed to decrypt data: ${error.message}`);
        }
    }

    async signData(data: Buffer, privateKey: string): Promise<Buffer> {
        this.ensureInitialized();

        try {
            return await this.core.securitySignData(data, privateKey);
        } catch (error) {
            throw new SatoxError(`Failed to sign data: ${error.message}`);
        }
    }

    async verifySignature(data: Buffer, signature: Buffer, publicKey: string): Promise<boolean> {
        this.ensureInitialized();

        try {
            return await this.core.securityVerifySignature(data, signature, publicKey);
        } catch (error) {
            throw new SatoxError(`Failed to verify signature: ${error.message}`);
        }
    }

    registerCallback(callback: (event: string, data: any) => void): void {
        this.callbacks.push(callback);
    }

    unregisterCallback(callback: (event: string, data: any) => void): void {
        const index = this.callbacks.indexOf(callback);
        if (index > -1) {
            this.callbacks.splice(index, 1);
        }
    }

    private ensureInitialized(): void {
        if (!this.initialized) {
            throw new SatoxError('SecurityManager not initialized');
        }
    }
} 